frontend/src/components/MSA.tsx

```diff
@@ -1,247 +1,292 @@
-  colors: manualColors = {},
+  colorMap: manualColors = {},
 }: Props) => {
-  const ref = useRef<HTMLDivElement | null>(null);
-
-  /** full width */
-  let [expanded, setExpanded] = useLocalStorage("msa-expanded", false);
-  expanded ??= false;
+  /** whether to wrap sequence to separate "panels" */
+  const [wrap, setWrap] = useState(true);
 
   /** maximum sequence length */
   const length = useMemo(
     () => max(tracks.map((track) => track.sequence.length)) ?? 0,
     [tracks],
   );
 
-  /** wrap sequence to new row if more than this num of chars */
-  const [wrap, setWrap] = useState(length);
-
   /** assign types */
   const { combinedWithTypes, tracksWithTypes, types } = useMemo(
     () => getDerived(tracks, length, getType),
     [tracks, length, getType],
   );
 
   /** map of type to color */
-  const colors = useColorMap(types, "mode", manualColors);
-
-  /** split sequence into chunks */
-  const chunks: [number, number][] = pairs(
-    range(0, length, wrap).concat([length]),
-  );
+  const colorMap = useColorMap(types, "mode", manualColors);
 
-  return (
-    <Flex direction="column" full>
-      <Flex
-        ref={ref}
-        direction="column"
-        hAlign="left"
-        className={clsx("card", classes.root, expanded && classes.expanded)}
-      >
-        {/* chunks */}
-        {chunks.map(([start, end], index) => (
-          <MSAChunk
-            key={index}
-            tracks={tracksWithTypes.map((track) => ({
-              ...track,
-              sequence: track.sequence.slice(start, end),
-            }))}
-            colors={colors}
-            combined={combinedWithTypes.slice(start, end)}
-            start={start}
-            end={end}
-          />
-        ))}
+  const theme = useTheme();
 
-        {/* legend */}
-        <Legend entries={mapValues(colors, (color) => ({ color }))} />
-      </Flex>
+  const truncateWidth = useTruncateWidth();
 
-      {/* controls */}
-      <Flex gap="lg" gapRatio={0.5}>
-        <NumberBox
+  return (
+    <Chart
+      title={title}
+      filename={[...filename, "msa"]}
+      full
+      controls={[
+        <CheckBox
           label="Wrap"
-          layout="horizontal"
           value={wrap}
-          onChange={setWrap}
-          min={10}
-          max={round(length, 10, "ceil")}
-          step={5}
-          tooltip="Wrap to new rows if sequence longer than this many characters"
-        />
+          onChange={(value, event) => {
+            if (event.currentTarget) preserveScroll(event.currentTarget);
+            setWrap(value);
+          }}
+          tooltip="Wrap sequence to stacked panels"
+        />,
+      ]}
+    >
+      {({ width }) => {
+        /** max num of chars that can fit in width */
+        const rowChars = wrap
+          ? Math.max(Math.floor((width - labelWidth) / charWidth), minChars)
+          : length;
+
+        /** split sequence into multiple panels */
+        const panels: [number, number][] = pairs(
```

Comment:
This code is nearly all the same, though it doesn't look like it from the diff. I previously had a separate "chunk" sub-component, but it is now just all in this render callback. The "chunks" are now called "panels", since I don't know what else to call them ("rows" already refers to each row of the viz).